# Nama : Muhammad Jauhar Hakim
# Data Scientist Application - PT ITSEC Asia Tbk

In [1]:
# Provide your credentials to the runtime
from google.colab import auth
from google.cloud import bigquery
auth.authenticate_user()
print('Authenticated')

Authenticated


In [2]:
project_id = 'belaaajarrr'

In [3]:
client = bigquery.Client(project=project_id)

# Citibike Station

In [4]:
sample_count = 5

df_citibike_station = client.query('''
SELECT *
FROM bigquery-public-data.new_york_citibike.citibike_stations
LIMIT %d
'''%(sample_count)).to_dataframe()

df_citibike_station

station_id                        name  \
0  47da6ab2-c841-4f2d-97ef-dd8d441c96b1  Loring Pl North & W 183 St   
1  7d5fa1f0-4069-4d5d-b735-5d500597394a       W Broadway & Watts St   
2  c1a4d909-0a00-475a-8e82-18ed13a4eb01    Whitehall St & Bridge St   
3  e85c3165-ad0a-44a3-9683-d8cf7575f465         Caton Ave & E 10 St   
4                   1815968237323833702  Brooklyn Ave & Beverley Rd   

  short_name   latitude  longitude  region_id   rental_methods  capacity  \
0    8555.05  40.860424 -73.909336         71  KEY, CREDITCARD        20   
1    5569.07  40.723230 -74.003140         71  KEY, CREDITCARD        51   
2    4962.02  40.703662 -74.013181         71  KEY, CREDITCARD        44   
3    3230.01  40.648481 -73.970580         71  KEY, CREDITCARD        30   
4    3164.04  40.645750 -73.943310          0  KEY, CREDITCARD        21   

   eightd_has_key_dispenser  num_bikes_available  num_bikes_disabled  \
0                     False                    0                   0   
1                     False                    0                   0   
2                     False                    0                   0   
3                     False                    0                   0   
4                     False                    0                   0   

   num_docks_available  num_docks_disabled  is_installed  is_renting  \
0                    0                   0         False       False   
1                    0                   0         False       False   
2                    0                   0         False       False   
3                    0                   0         False       False   
4                    0                   0         False       False   

   is_returning  eightd_has_available_keys             last_reported  
0         False                      False 2024-08-06 12:16:31+00:00  
1         False                      False 2024-05-31 16:22:42+00:00  
2         False                      False 2024-07-11 18:05:28+00:00  
3         False                      False 2024-07-17 14:01:29+00:00  
4         False                      False 2024-08-13 14:05:49+00:00

In [5]:
df_citibike_station.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 18 columns):
 #   Column                     Non-Null Count  Dtype              
---  ------                     --------------  -----              
 0   station_id                 5 non-null      object             
 1   name                       5 non-null      object             
 2   short_name                 5 non-null      object             
 3   latitude                   5 non-null      float64            
 4   longitude                  5 non-null      float64            
 5   region_id                  5 non-null      Int64              
 6   rental_methods             5 non-null      object             
 7   capacity                   5 non-null      Int64              
 8   eightd_has_key_dispenser   5 non-null      boolean            
 9   num_bikes_available        5 non-null      Int64              
 10  num_bikes_disabled         5 non-null      Int64              
 11  num_docks_

# Citibike Trips

In [6]:
df_citibike_trips = client.query('''
SELECT *
FROM bigquery-public-data.new_york_citibike.citibike_trips
WHERE start_station_id IS NOT NULL
LIMIT %d
'''%(sample_count)).to_dataframe()

df_citibike_trips

tripduration           starttime            stoptime  start_station_id  \
0           166 2017-10-02 14:39:41 2017-10-02 14:42:28              3480   
1          1614 2016-06-30 17:19:24 2016-06-30 17:46:18               520   
2           655 2017-07-15 06:57:36 2017-07-15 07:08:31               520   
3           226 2017-08-04 08:31:55 2017-08-04 08:35:41               520   
4           261 2017-07-18 18:32:41 2017-07-18 18:37:02               520   

  start_station_name  start_station_latitude  start_station_longitude  \
0       WS Don't Use                0.000000                 0.000000   
1    W 52 St & 5 Ave               40.759923               -73.976485   
2    W 52 St & 5 Ave               40.759923               -73.976485   
3    W 52 St & 5 Ave               40.759923               -73.976485   
4    W 52 St & 5 Ave               40.759923               -73.976485   

   end_station_id end_station_name  end_station_latitude  \
0            3187        Warren St             40.721124   
1            3122    48 Ave & 5 St             40.744363   
2            3459  E 53 St & 3 Ave             40.757632   
3            3459  E 53 St & 3 Ave             40.757632   
4            3459  E 53 St & 3 Ave             40.757632   

   end_station_longitude  bikeid    usertype  birth_year   gender  \
0             -74.038051   31952  Subscriber        1980     male   
1             -73.955873   25747  Subscriber        1973     male   
2             -73.969306   17133    Customer        <NA>  unknown   
3             -73.969306   15372  Subscriber        1965     male   
4             -73.969306   28624  Subscriber        1980     male   

  customer_plan  
0                
1                
2                
3                
4

In [7]:
df_citibike_trips.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 16 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   tripduration             5 non-null      Int64         
 1   starttime                5 non-null      datetime64[us]
 2   stoptime                 5 non-null      datetime64[us]
 3   start_station_id         5 non-null      Int64         
 4   start_station_name       5 non-null      object        
 5   start_station_latitude   5 non-null      float64       
 6   start_station_longitude  5 non-null      float64       
 7   end_station_id           5 non-null      Int64         
 8   end_station_name         5 non-null      object        
 9   end_station_latitude     5 non-null      float64       
 10  end_station_longitude    5 non-null      float64       
 11  bikeid                   5 non-null      Int64         
 12  usertype                 5 non-null     

# Task 1

*   What proportion in percentage of remaining trips terminate at end_station_id that likewise do not exist in the station table?
*   You should eliminate the journeys from the trip table that lack a start_station_id, in remaining trips, keep those with start_station_id that were note present at the station table.



In [8]:
df_task1 = client.query('''
  SELECT COUNT(*) AS Remaining_Trips
  FROM (SELECT *
        FROM bigquery-public-data.new_york_citibike.citibike_trips
        WHERE start_station_id IS NOT NULL) AS CT
  INNER JOIN bigquery-public-data.new_york_citibike.citibike_stations AS CS
  ON CAST(CT.end_station_id AS STRING) = CAST(CS.station_id AS STRING);
''').to_dataframe()
df_task1

Remaining_Trips
0                0

In [9]:
df_task11 = client.query('''
  SELECT CT.*, CS.*
  FROM (SELECT *
        FROM bigquery-public-data.new_york_citibike.citibike_trips
        WHERE start_station_id IS NOT NULL) AS CT
  INNER JOIN bigquery-public-data.new_york_citibike.citibike_stations AS CS
  ON CAST(CT.end_station_name AS STRING) = CAST(CS.name AS STRING)
  LIMIT 100;
''').to_dataframe()
df_task11

tripduration               starttime                stoptime  \
0            166 2017-10-02 14:39:41.000 2017-10-02 14:42:28.000   
1           1614 2016-06-30 17:19:24.000 2016-06-30 17:46:18.000   
2            655 2017-07-15 06:57:36.000 2017-07-15 07:08:31.000   
3            226 2017-08-04 08:31:55.000 2017-08-04 08:35:41.000   
4            261 2017-07-18 18:32:41.000 2017-07-18 18:37:02.000   
..           ...                     ...                     ...   
95           529 2017-10-21 13:34:15.000 2017-10-21 13:43:04.000   
96           974 2018-05-02 17:06:10.578 2018-05-02 17:22:25.009   
97           954 2018-03-28 16:34:10.021 2018-03-28 16:50:04.748   
98           876 2017-10-11 17:35:47.000 2017-10-11 17:50:23.000   
99          1422 2017-09-15 14:54:58.000 2017-09-15 15:18:40.000   

    start_station_id start_station_name  start_station_latitude  \
0               3480       WS Don't Use                0.000000   
1                520    W 52 St & 5 Ave               40.759923   
2                520    W 52 St & 5 Ave               40.759923   
3                520    W 52 St & 5 Ave               40.759923   
4                520    W 52 St & 5 Ave               40.759923   
..               ...                ...                     ...   
95               520    W 52 St & 5 Ave               40.759923   
96               520    W 52 St & 5 Ave               40.759923   
97               520    W 52 St & 5 Ave               40.759923   
98               520    W 52 St & 5 Ave               40.759923   
99               520    W 52 St & 5 Ave               40.759923   

    start_station_longitude  end_station_id          end_station_name  \
0                  0.000000            3187                 Warren St   
1                -73.976485            3122             48 Ave & 5 St   
2                -73.976485            3459           E 53 St & 3 Ave   
3                -73.976485            3459           E 53 St & 3 Ave   
4                -73.976485            3459           E 53 St & 3 Ave   
..                      ...             ...                       ...   
95               -73.976485            3255           8 Ave & W 31 St   
96               -73.976485            3255           8 Ave & W 31 St   
97               -73.976485            3255           8 Ave & W 31 St   
98               -73.976485            3255           8 Ave & W 31 St   
99               -73.976485             151  Cleveland Pl & Spring St   

    end_station_latitude  ...  eightd_has_key_dispenser  num_bikes_available  \
0              40.721124  ...                     False                   15   
1              40.744363  ...                     False                   43   
2              40.757632  ...                     False                    0   
3              40.757632  ...                     False                    0   
4              40.757632  ...                     False                    0   
..                   ...  ...                       ...                  ...   
95             40.750585  ...                     False                    6   
96             40.750585  ...                     False                    6   
97             40.750585  ...                     False                    6   
98             40.750585  ...                     False                    6   
99             40.722104  ...                     False                    7   

   num_bikes_disabled  num_docks_available num_docks_disabled is_installed  \
0                   0                    6                  0         True   
1                   5                   27                  0         True   
2                   0                   41                  0         True   
3                   0                   41                  0         True   
4                   0                   41                  0         True   
..                ...                  ...                ...          ...  

In [10]:
df_task12 = client.query('''
  SELECT COUNT(*) AS Remaining_Trips
  FROM (SELECT *
        FROM bigquery-public-data.new_york_citibike.citibike_trips
        WHERE start_station_id IS NOT NULL) AS CT
  WHERE CT.end_station_name NOT IN (SELECT name FROM bigquery-public-data.new_york_citibike.citibike_stations);
''').to_dataframe()
df_task12

Remaining_Trips
0         13074073

In [11]:
df_task13 = client.query('''
  SELECT COUNT(*) AS Total_Trips
  FROM bigquery-public-data.new_york_citibike.citibike_trips
  WHERE start_station_id IS NOT NULL;
''').to_dataframe()
df_task13

Total_Trips
0     53108721

In [12]:
df_task14 = client.query('''
SELECT  ttable.Total_Trips,
        rtable.Remaining_Trips,
        CONCAT(CAST(ROUND(rtable.Remaining_Trips*100/ttable.Total_Trips, 2) AS STRING), '%') AS Percentage_Remaining_Trips
FROM (
    SELECT COUNT(*) AS Total_Trips
    FROM bigquery-public-data.new_york_citibike.citibike_trips
    WHERE start_station_id IS NOT NULL
    ) AS ttable
CROSS JOIN (
    SELECT COUNT(*) AS Remaining_Trips
    FROM (
      SELECT *
      FROM bigquery-public-data.new_york_citibike.citibike_trips
      WHERE start_station_id IS NOT NULL) AS CT
    WHERE CT.end_station_name NOT IN (SELECT name FROM bigquery-public-data.new_york_citibike.citibike_stations)
    ) AS rtable
''').to_dataframe()
df_task14

Total_Trips  Remaining_Trips Percentage_Remaining_Trips
0     53108721         13074073                     24.62%

**We can conclude the percentage of remaining trips terminate at end_station_id that likewise do not exist in the station table around 24.62%**


# Task 2

*   Retain only those records in the trip table which have a starttime from 1st January 2018 onwards.
*   Merge the usertype, birth_year, and gender fields into a single entity, considering each unique combination as one user. Include those users who have missing data for usertype, birth_year, or gender.

For each month, categorize users into groups based on the data from their trips that month:
*   If the user has 0 unique start_station_name, classify them as "inactive."
*   If the user has between 1 and 10 unique start_station_name, label them as "casual.”
*   If the user has more than 10 unique start_station_name, label them as "power."

It's crucial to fill in missing month data with 0. For instance, if user A has data for the 1st and 3rd months but lacks data for the 2nd month, then the 2nd month must be filled with 0, and hence user A would be classified as "inactive" for that month.

Here are your questions:

a. In each month of 2018, what is the user count in each segment?

b. For each month of 2018, determine the shift of users among the segments in the following month. As an example: From January 2018 to February 2018, quantify how many casual users remained casual, transitioned to power, or became inactive? Repeat this process for the other categories and for the rest of the months in 2018.








In [13]:
df_task21 = client.query('''
  SELECT A.Month_Year_Trip, B.User
  FROM (SELECT DISTINCT FORMAT_DATE('%m-%Y', starttime) AS Month_Year_Trip
        FROM bigquery-public-data.new_york_citibike.citibike_trips
        WHERE starttime>'2018-01-01') A
  CROSS JOIN (SELECT DISTINCT CONCAT(COALESCE(CAST(usertype AS STRING),"NULL"),
                '-', COALESCE(CAST(birth_year AS STRING),"NULL"),
                '-', COALESCE(CAST(gender AS STRING),"NULL")) AS User
              FROM bigquery-public-data.new_york_citibike.citibike_trips
              WHERE starttime>'2018-01-01') B;
''').to_dataframe()
df_task21

Month_Year_Trip                     User
0            01-2018    Customer-1897-unknown
1            01-2018     Customer-1972-female
2            01-2018     Customer-1949-female
3            01-2018       Customer-1941-male
4            01-2018   Subscriber-1987-female
...              ...                      ...
2155         05-2018    Customer-1974-unknown
2156         05-2018     Subscriber-1971-male
2157         05-2018   Subscriber-1992-female
2158         05-2018    Customer-1991-unknown
2159         05-2018  Subscriber-1885-unknown

[2160 rows x 2 columns]

In [14]:
df_task22 = client.query('''
  WITH all_combinations AS (
    SELECT A.Month_Year_Trip, B.User
    FROM (SELECT DISTINCT FORMAT_DATE('%m-%Y', starttime) AS Month_Year_Trip
          FROM bigquery-public-data.new_york_citibike.citibike_trips
          WHERE starttime>'2018-01-01') A
    CROSS JOIN (SELECT DISTINCT CONCAT(COALESCE(CAST(usertype AS STRING),"NULL"),
                  '-', COALESCE(CAST(birth_year AS STRING),"NULL"),
                  '-', COALESCE(CAST(gender AS STRING),"NULL")) AS User
                FROM bigquery-public-data.new_york_citibike.citibike_trips
                WHERE starttime>'2018-01-01') B
  )
  SELECT AC.Month_Year_Trip, AC.USER,
         COUNT(DISTINCT start_station_name) AS Combination_Count,
         CASE
            WHEN COUNT(DISTINCT start_station_name) = 0 THEN 'inactive'
            WHEN COUNT(DISTINCT start_station_name) > 0 AND COUNT(DISTINCT start_station_name)<= 10 THEN 'casual'
            ELSE 'power'
         END AS Group_Category
  FROM all_combinations AC
  LEFT JOIN bigquery-public-data.new_york_citibike.citibike_trips CT
    ON AC.Month_Year_Trip = FORMAT_DATE('%m-%Y', CT.starttime)
    AND AC.User = CONCAT(COALESCE(CAST(CT.usertype AS STRING),"NULL"),
                        '-', COALESCE(CAST(CT.birth_year AS STRING),"NULL"),
                        '-', COALESCE(CAST(CT.gender AS STRING),"NULL"))
  GROUP BY AC.Month_Year_Trip, AC.User;
''').to_dataframe()
df_task22

Month_Year_Trip                     USER  Combination_Count  \
0            01-2018       Customer-1971-male                 41   
1            02-2018     Subscriber-1963-male                582   
2            04-2018     Customer-1995-female                365   
3            01-2018       Customer-1957-male                  9   
4            03-2018  Subscriber-1984-unknown                 97   
...              ...                      ...                ...   
2155         01-2018    Customer-1954-unknown                  0   
2156         01-2018   Subscriber-1931-female                  0   
2157         01-2018       Customer-1924-male                  0   
2158         05-2018     Subscriber-1915-male                  1   
2159         03-2018  Subscriber-1930-unknown                  0   

     Group_Category  
0             power  
1             power  
2             power  
3            casual  
4             power  
...             ...  
2155       inactive  
2156       inactive  
2157       inactive  
2158         casual  
2159       inactive  

[2160 rows x 4 columns]

In [15]:
df_task23 = client.query('''
WITH search_each_segment AS (
    WITH all_combinations AS (
      SELECT A.Month_Year_Trip, B.User
      FROM (SELECT DISTINCT FORMAT_DATE('%m-%Y', starttime) AS Month_Year_Trip
            FROM bigquery-public-data.new_york_citibike.citibike_trips
            WHERE starttime>'2018-01-01') A
      CROSS JOIN (SELECT DISTINCT CONCAT(COALESCE(CAST(usertype AS STRING),"NULL"),
                    '-', COALESCE(CAST(birth_year AS STRING),"NULL"),
                    '-', COALESCE(CAST(gender AS STRING),"NULL")) AS User
                  FROM bigquery-public-data.new_york_citibike.citibike_trips
                  WHERE starttime>'2018-01-01') B
    )
    SELECT AC.Month_Year_Trip, AC.USER,
          COUNT(DISTINCT start_station_name) AS Combination_Count,
          CASE
              WHEN COUNT(DISTINCT start_station_name) = 0 THEN 'inactive'
              WHEN COUNT(DISTINCT start_station_name) > 0 AND COUNT(DISTINCT start_station_name)<= 10 THEN 'casual'
              ELSE 'power'
          END AS Group_Category
    FROM all_combinations AC
    LEFT JOIN bigquery-public-data.new_york_citibike.citibike_trips CT
      ON AC.Month_Year_Trip = FORMAT_DATE('%m-%Y', CT.starttime)
      AND AC.User = CONCAT(COALESCE(CAST(CT.usertype AS STRING),"NULL"),
                          '-', COALESCE(CAST(CT.birth_year AS STRING),"NULL"),
                          '-', COALESCE(CAST(CT.gender AS STRING),"NULL"))
    GROUP BY AC.Month_Year_Trip, AC.User
    )
  SELECT SES.Month_Year_Trip, SES.Group_Category, COUNT(SES.Group_Category) AS User_Count
  FROM search_each_segment SES
  GROUP BY SES.Month_Year_Trip,SES.Group_Category
  ORDER BY Month_Year_Trip;
''').to_dataframe()
df_task23

Month_Year_Trip Group_Category  User_Count
0          01-2018          power         249
1          01-2018         casual          68
2          01-2018       inactive         115
3          02-2018          power         265
4          02-2018         casual          80
5          02-2018       inactive          87
6          03-2018          power         276
7          03-2018         casual          74
8          03-2018       inactive          82
9          04-2018          power         308
10         04-2018         casual          87
11         04-2018       inactive          37
12         05-2018          power         328
13         05-2018         casual          83
14         05-2018       inactive          21

**We can conclude the user count per category per month in 2018 (the data only exist from January to May)**

In [16]:
df_task24 = client.query('''
  WITH all_combinations AS (
    SELECT A.Month_Year_Trip, B.User
    FROM (SELECT DISTINCT FORMAT_DATE('%m-%Y', starttime) AS Month_Year_Trip
          FROM bigquery-public-data.new_york_citibike.citibike_trips
          WHERE starttime>'2018-01-01') A
    CROSS JOIN (SELECT DISTINCT CONCAT(COALESCE(CAST(usertype AS STRING),"NULL"),
                  '-', COALESCE(CAST(birth_year AS STRING),"NULL"),
                  '-', COALESCE(CAST(gender AS STRING),"NULL")) AS User
                FROM bigquery-public-data.new_york_citibike.citibike_trips
                WHERE starttime>'2018-01-01') B
  ), previous_month_combination AS (
    SELECT AC.Month_Year_Trip, AC.USER,
           FORMAT_DATE('%m-%Y', DATE_SUB(PARSE_DATE('%m-%Y', AC.Month_Year_Trip), INTERVAL 1 MONTH)) AS Previous_Month_Year_Trip,
           COUNT(DISTINCT CT.start_station_name) AS Previous_Combination_Count,
           CASE
            WHEN COUNT(DISTINCT CT.start_station_name) = 0 THEN 'inactive'
            WHEN COUNT(DISTINCT CT.start_station_name) > 0 AND COUNT(DISTINCT CT.start_station_name)<= 10 THEN 'casual'
            ELSE 'power'
           END AS Previous_Group_Category
    FROM all_combinations AC
    LEFT JOIN bigquery-public-data.new_york_citibike.citibike_trips CT
      ON FORMAT_DATE('%m-%Y', DATE_SUB(PARSE_DATE('%m-%Y', AC.Month_Year_Trip), INTERVAL 1 MONTH)) = FORMAT_DATE('%m-%Y', CT.starttime)
      AND AC.User = CONCAT(COALESCE(CAST(CT.usertype AS STRING),"NULL"),
                          '-', COALESCE(CAST(CT.birth_year AS STRING),"NULL"),
                          '-', COALESCE(CAST(CT.gender AS STRING),"NULL"))
    GROUP BY AC.Month_Year_Trip, AC.User
  ), current_month_combination AS (
    SELECT AC.Month_Year_Trip, AC.USER,
           COUNT(DISTINCT start_station_name) AS Combination_Count,
           CASE
            WHEN COUNT(DISTINCT start_station_name) = 0 THEN 'inactive'
            WHEN COUNT(DISTINCT start_station_name) > 0 AND COUNT(DISTINCT start_station_name)<= 10 THEN 'casual'
            ELSE 'power'
           END AS Group_Category
    FROM all_combinations AC
    LEFT JOIN bigquery-public-data.new_york_citibike.citibike_trips CT
      ON AC.Month_Year_Trip = FORMAT_DATE('%m-%Y', CT.starttime)
      AND AC.User = CONCAT(COALESCE(CAST(CT.usertype AS STRING),"NULL"),
                          '-', COALESCE(CAST(CT.birth_year AS STRING),"NULL"),
                          '-', COALESCE(CAST(CT.gender AS STRING),"NULL"))
    GROUP BY AC.Month_Year_Trip, AC.User
  )
SELECT  CMC.*,
        PMC.Previous_Month_Year_Trip, PMC.Previous_Combination_Count, PMC.Previous_Group_Category
FROM current_month_combination CMC
LEFT JOIN previous_month_combination PMC
  ON CMC.Month_Year_Trip = PMC.Month_Year_Trip
  AND CMC.User = PMC.User;
''').to_dataframe()
df_task24

Month_Year_Trip                    USER  Combination_Count  \
0            04-2018    Customer-1963-female                 89   
1            01-2018    Subscriber-1962-male                559   
2            04-2018   Customer-1987-unknown                 32   
3            01-2018      Customer-1971-male                 41   
4            02-2018    Subscriber-1963-male                582   
...              ...                     ...                ...   
2155         02-2018  Subscriber-1936-female                  0   
2156         02-2018      Customer-1925-male                  0   
2157         01-2018    Subscriber-1933-male                  0   
2158         01-2018    Subscriber-1887-male                  0   
2159         01-2018  Subscriber-1901-female                  0   

     Group_Category Previous_Month_Year_Trip  Previous_Combination_Count  \
0             power                  03-2018                          24   
1             power                  12-2017                         584   
2             power                  03-2018                          36   
3             power                  12-2017                          52   
4             power                  01-2018                         583   
...             ...                      ...                         ...   
2155       inactive                  01-2018                           1   
2156       inactive                  01-2018                           0   
2157       inactive                  12-2017                           0   
2158       inactive                  12-2017                           1   
2159       inactive                  12-2017                           0   

     Previous_Group_Category  
0                      power  
1                      power  
2                      power  
3                      power  
4                      power  
...                      ...  
2155                  casual  
2156                inactive  
2157                inactive  
2158                  casual  
2159                inactive  

[2160 rows x 7 columns]

In [17]:
df_task25 = client.query('''
  WITH all_combinations AS (
    SELECT A.Month_Year_Trip, B.User
    FROM (SELECT DISTINCT FORMAT_DATE('%m-%Y', starttime) AS Month_Year_Trip
          FROM bigquery-public-data.new_york_citibike.citibike_trips
          WHERE starttime>'2018-01-01') A
    CROSS JOIN (SELECT DISTINCT CONCAT(COALESCE(CAST(usertype AS STRING),"NULL"),
                  '-', COALESCE(CAST(birth_year AS STRING),"NULL"),
                  '-', COALESCE(CAST(gender AS STRING),"NULL")) AS User
                FROM bigquery-public-data.new_york_citibike.citibike_trips
                WHERE starttime>'2018-01-01') B
  ), previous_month_combination AS (
    SELECT AC.Month_Year_Trip, AC.USER,
           FORMAT_DATE('%m-%Y', DATE_SUB(PARSE_DATE('%m-%Y', AC.Month_Year_Trip), INTERVAL 1 MONTH)) AS Previous_Month_Year_Trip,
           COUNT(DISTINCT CT.start_station_name) AS Previous_Combination_Count,
           CASE
            WHEN COUNT(DISTINCT CT.start_station_name) = 0 THEN 'inactive'
            WHEN COUNT(DISTINCT CT.start_station_name) > 0 AND COUNT(DISTINCT CT.start_station_name)<= 10 THEN 'casual'
            ELSE 'power'
           END AS Previous_Group_Category
    FROM all_combinations AC
    LEFT JOIN bigquery-public-data.new_york_citibike.citibike_trips CT
      ON FORMAT_DATE('%m-%Y', DATE_SUB(PARSE_DATE('%m-%Y', AC.Month_Year_Trip), INTERVAL 1 MONTH)) = FORMAT_DATE('%m-%Y', CT.starttime)
      AND AC.User = CONCAT(COALESCE(CAST(CT.usertype AS STRING),"NULL"),
                          '-', COALESCE(CAST(CT.birth_year AS STRING),"NULL"),
                          '-', COALESCE(CAST(CT.gender AS STRING),"NULL"))
    GROUP BY AC.Month_Year_Trip, AC.User
  ), current_month_combination AS (
    SELECT AC.Month_Year_Trip, AC.USER,
           COUNT(DISTINCT start_station_name) AS Combination_Count,
           CASE
            WHEN COUNT(DISTINCT start_station_name) = 0 THEN 'inactive'
            WHEN COUNT(DISTINCT start_station_name) > 0 AND COUNT(DISTINCT start_station_name)<= 10 THEN 'casual'
            ELSE 'power'
           END AS Group_Category
    FROM all_combinations AC
    LEFT JOIN bigquery-public-data.new_york_citibike.citibike_trips CT
      ON AC.Month_Year_Trip = FORMAT_DATE('%m-%Y', CT.starttime)
      AND AC.User = CONCAT(COALESCE(CAST(CT.usertype AS STRING),"NULL"),
                          '-', COALESCE(CAST(CT.birth_year AS STRING),"NULL"),
                          '-', COALESCE(CAST(CT.gender AS STRING),"NULL"))
    GROUP BY AC.Month_Year_Trip, AC.User
  )
SELECT  CMC.Month_Year_Trip, CMC.Group_Category,
        PMC.Previous_Group_Category, COUNT(Group_Category) AS User_Count
FROM current_month_combination CMC
LEFT JOIN previous_month_combination PMC
  ON CMC.Month_Year_Trip = PMC.Month_Year_Trip
  AND CMC.User = PMC.User
GROUP BY CMC.Month_Year_Trip, CMC.Group_Category, PMC.Previous_Group_Category
ORDER BY CMC.Month_Year_Trip, CMC.Group_Category;
''').to_dataframe()
df_task25

Month_Year_Trip Group_Category Previous_Group_Category  User_Count
0          01-2018         casual                   power          18
1          01-2018         casual                  casual          37
2          01-2018         casual                inactive          13
3          01-2018       inactive                  casual          27
4          01-2018       inactive                inactive          86
5          01-2018       inactive                   power           2
6          01-2018          power                   power         242
7          01-2018          power                  casual           4
8          01-2018          power                inactive           3
9          02-2018         casual                  casual          37
10         02-2018         casual                inactive          36
11         02-2018         casual                   power           7
12         02-2018       inactive                inactive          76
13         02-2018       inactive                  casual           9
14         02-2018       inactive                   power           2
15         02-2018          power                   power         240
16         02-2018          power                  casual          22
17         02-2018          power                inactive           3
18         03-2018         casual                   power          12
19         03-2018         casual                inactive          17
20         03-2018         casual                  casual          45
21         03-2018       inactive                inactive          66
22         03-2018       inactive                  casual          16
23         03-2018          power                   power         253
24         03-2018          power                  casual          19
25         03-2018          power                inactive           4
26         04-2018         casual                  casual          38
27         04-2018         casual                inactive          42
28         04-2018         casual                   power           7
29         04-2018       inactive                inactive          32
30         04-2018       inactive                  casual           5
31         04-2018          power                   power         269
32         04-2018          power                  casual          31
33         04-2018          power                inactive           8
34         05-2018         casual                  casual          45
35         05-2018         casual                   power           7
36         05-2018         casual                inactive          31
37         05-2018       inactive                inactive           5
38         05-2018       inactive                  casual          16
39         05-2018          power                   power         301
40         05-2018          power                  casual          26
41         05-2018          power                inactive           1

**We can conclude the user count per category per month in 2018 (the data only exist from January to May) that shifting category from previous month**

# Task 3


*   Please write a query that generates the name of each credit card along with the discrepancy in the number of cards issued between the month with the maximum issuance and the month with minimum issuance
*   Arrange the outcome in descending order



In [23]:
df_task31 = client.query('''
  SELECT *
  FROM belaaajarrr.test_dataset.Issued_monthly_cards
''').to_dataframe()
df_task31

name_card  ammount_issued  month_issue  year_issue
0    Card A           55000            1        2021
1    Card A           60000            2        2021
2    Card A           65000            3        2021
3    Card A           70000            4        2021
4    Card B          170000            1        2021
5    Card B          175000            2        2021
6    Card B          180000            3        2021
7    Card C           10000            2        2021
8    Card C           90000            3        2021
9    Card C          180000            4        2021

In [22]:
df_task32 = client.query('''
  SELECT  name_card AS card_name,
          MAX(ammount_issued) - MIN(ammount_issued) AS difference
  FROM belaaajarrr.test_dataset.Issued_monthly_cards
  GROUP BY name_card;
''').to_dataframe()
df_task32

card_name  difference
0    Card A       15000
1    Card B       10000
2    Card C      170000

**We can conclude the max different of ammount issued by Card C around 170000**

# Task 4


*   How many customers, each month, make more than one deposit and at least one transaction (either a purchase or a withdrawal) within the same month?



In [37]:
df_task41 = client.query('''
  SELECT *
  FROM belaaajarrr.test_dataset.customer_transaction
''').to_dataframe()
df_task41

customer_id     tx_date     tx_type  tx_amount
0           429  2020-01-21     deposit         82
1           398  2020-01-01     deposit        196
2           185  2020-01-29     deposit        626
3           309  2020-01-13     deposit        995
4           376  2020-01-03     deposit        706
5           188  2020-01-13     deposit        601
6           429  2020-01-25     deposit        104
7           429  2020-01-27     deposit         55
8           429  2020-01-29     deposit        155
9           185  2020-01-13     deposit        400
10          155  2020-01-10    purchase        712
11          255  2020-01-14    purchase        563
12          312  2020-01-20  withdrawal        485
13          138  2020-01-11  withdrawal        520
14          185  2020-01-17  withdrawal        250

In [38]:
df_task42 = client.query('''
  SELECT
      customer_id,
      FORMAT_DATE('%m-%Y', tx_date) AS month,
      COUNT(*) AS deposit_count
  FROM belaaajarrr.test_dataset.customer_transaction
  WHERE tx_type = 'deposit'
  GROUP BY customer_id, month
''').to_dataframe()
df_task42

customer_id    month  deposit_count
0          429  01-2020              4
1          398  01-2020              1
2          185  01-2020              2
3          309  01-2020              1
4          376  01-2020              1
5          188  01-2020              1

In [39]:
df_task43 = client.query('''
  SELECT
      customer_id,
      FORMAT_DATE('%m-%Y', tx_date) AS month
  FROM belaaajarrr.test_dataset.customer_transaction
  WHERE tx_type IN ('purchase', 'withdrawal')
  GROUP BY customer_id, month
''').to_dataframe()
df_task43

customer_id    month
0          155  01-2020
1          255  01-2020
2          312  01-2020
3          138  01-2020
4          185  01-2020

In [40]:
df_task44 = client.query('''
WITH monthly_deposits AS (
    SELECT
        customer_id,
        FORMAT_DATE('%m-%Y', tx_date) AS month,
        COUNT(*) AS deposit_count
    FROM belaaajarrr.test_dataset.customer_transaction
    WHERE tx_type = 'deposit'
    GROUP BY customer_id, month
),
monthly_other_transactions AS (
    SELECT
        customer_id,
        FORMAT_DATE('%m-%Y', tx_date) AS month
    FROM belaaajarrr.test_dataset.customer_transaction
    WHERE tx_type IN ('purchase', 'withdrawal')
    GROUP BY customer_id, month
),
eligible_customers AS (
    SELECT
        md.customer_id,
        md.month
    FROM monthly_deposits md
    JOIN monthly_other_transactions mot
    ON md.customer_id = mot.customer_id AND md.month = mot.month
    WHERE md.deposit_count > 1
)
SELECT
    month,
    COUNT(DISTINCT customer_id) AS customer_count
FROM eligible_customers
GROUP BY month
ORDER BY month;
''').to_dataframe()
df_task44

month  customer_count
0  01-2020               1

**We can conclude in January 2020 who have more than one deposit and at least one purchase or one withdrawal is around 1 customer**